vcmi/vcmi
# Patch III for Mod system improvement : Negative bonuses: new feature, fixes, town building: Magic Well

- [x] New feature: Negative town bonuses for enemy army implementation
- [x] Fix: Negative bonus bearers should not be affected by own negative bonuses. So, devils, archdevils, negative artifacts (shaman's thing and some dead head), negative hero's abilities (such as Amanda from  Cathedral) should work now.
- [x] Fix: Visiting hero should receive town bonuses during town siege when he defends a town as garrison hero. Bonus tree mounting is fixed to provide town bonuses.

- [x] New feature: 'Magic Well' should work in fan towns.
- [x] New feature: Introduced new limiter. The advantage of this solution is the bonus propagation is implemented using a standard propagation mechanic, it needs just owner adjustment instead of whole bonus copying to the node on battle setup

**Samples:**

Magic Well: `"onVisitBonuses" : [ { "type": "TOWN_MAGIC_WELL" } ],`
Magic Absorber: `"bonuses" : [ { "type": "SPELL_DAMAGE_REDUCTION", "val": 50, "subtype": -1 } ]`
-2 morale to enemy army: `"bonuses" : [ { "type": "MORALE", "val": -2, "effectRange":"ONLY_ENEMY_ARMY" } ]`

**Guide to feature applying:**

No limiters or propagators  should be used with  ` "effectRange": "ONLY_ENEMY_ARMY" `

**Testing:**

Precondition: Devils are in the hero's army.

- Attack neutrals with devils outside a town.  
- Make sure, negatives are propagated from hero's army to the neutrals units.
- Make sure, own negatives are not propagated to the negative bonus owner.
- Attack neutral devils.  
- Make sure, negatives are propagated from neutrals units to the hero's army.
- Attack neutral devils with devils. 
- Make sure, enemy negatives are propagated to all units
- Attack neutral town (no devils in the hero's army) with devils in the town garrison. 
- Make sure, enemy negatives are propagated  from garrison to hero's army
- Attack neutral town (devils are the hero's army) with devils in the town garrison. 
- Make sure, enemy negatives are propagated to all units
- Attack town with built negative bonus building
- Make sure, negatives are propagated from town building to the hero's army
- Capture this town and remain weak  visitor (not garrison) hero as defender. Enemy should attack.
- Make sure, defending visitor obtains positive bonuses from town. 
- Make sure, enemy attacker obtains negative bonuses from town
- Win this battle. Move hero to garrison and wait for next attack.
- Make sure, defending garrison hero obtains positive bonuses from town. 
- Lose this battle and wait until enemy remains this town without guardians.
- Capture this town again, wait for enemy attack 
- Make sure, that negative bonuses from town work properly against enemy attacker only
- Get negative artifacts (shaman's thing, dead head and horseshoe), attack enemy neutral
- Make sure, that negative bonuses from artifacts work against neutrals only. Win battle
- Attack enemy hero. Make sure, that negative bonuses from artifacts work against enemy army only. 
- Choose hero with negative ability, attack enemy with this hero.
- Make sure, negative hero ability work properly against neutrals, enemy armies, town garrisons.
- Make sure, opposite bonuses without value (such as NO_MORALE) work properly against enemy in / outside town battles
- Make sure, that negative primary skills artifacts don't cause primary skill value to be less than valid minimum (0 and 1 for SP and Knowledge)
- Make sure Magic Well works properly (one time in day per hero)
- Make sure, negative artifacts work properly in the old saves
- Make sure, creatures with negative bonuses work properly in the old saves
- Make sure, negative hero's abilities work properly in the old saves
- Make sure, town buildings with  negative bonuses work properly in the old saves
